### ISSUE TITLE
C1: Typed operation semantics Implementation Plan

### ISSUE THREAD

**Issue by JueonPark** [JueonPark/sym#141]
**Goal:** Define and verify explicit cast, quantize, and dequantize operations for #132, with numerical semantics that the compiler, runtime, and frontend can test against one contract.

**Architecture:** Extend the reloc dialect with value-transform operations that preserve logical shape while declaring source/destination types and numerical policy. Keep value conversion separate from address mapping. Define the operation semantics first; existing kernels qualify only for the subset they actually implement.

**Tech Stack:** Existing LLVM/MLIR 21.1.8 and C++ settings. Numerical comparison with the qualified CPython 3.14.7 standard-GIL / PyTorch 2.14.0 CPU and cu126 environments; retain the current pybind11 3.0.4 bridge baseline.

**Spec:** [#132](https://github.com/JueonPark/sym/issues/132), [project plan §2/C1](https://github.com/JueonPark/sym/blob/main/docs/project-finalization-plan.md#2-extend-plans-with-cast-quantize-and-dequantize), and [qualification baseline](https://github.com/JueonPark/sym/blob/main/docs/torch-finalization/README.md#selected-version-baseline).

## Global Constraints

- "Preserve the CPU relocation plus pipelined transfer implementation and the existing compiler/runtime separation."
- "Any automatic placement must choose among implementations of the same requested computation; it must not silently introduce a cast or quantization to improve transfer time."
- Initial operations are f32→f16 and f16→f32 casts, f32→signed-i8 quantization, and signed-i8→f32 dequantization. Other type pairs require an explicit semantic/kernel extension.
- Quantization parameters come from the program or explicit configuration. No calibration, training, automatic lower-precision selection, or int4 packing is introduced.
- Layout inversion does not invert a lossy value transform. A D2H computation retains its requested forward operation order.
- Core libreloc remains MLIR/LLVM/Torch-free. Typed IR validity is separate from executable runtime capability.

---

**Proposed child title:** `[Finalize][Compiler][C1] Define typed relocation operation semantics`

**Dependencies:** Can start alongside the remaining layout-only integration. [C2](https://github.com/JueonPark/sym/issues/142) and [C3](https://github.com/JueonPark/sym/issues/143) consume this contract; R3 in [#133](https://github.com/JueonPark/sym/issues/133) supplies executable implementations, and [C4](https://github.com/JueonPark/sym/issues/144) checks them against it. [C1](https://github.com/JueonPark/sym/issues/141) does not wait for R3 or enable typed frontend execution.

**Compiler subissues:** [C1](https://github.com/JueonPark/sym/issues/141), [C2](https://github.com/JueonPark/sym/issues/142), [C3](https://github.com/JueonPark/sym/issues/143), [C4](https://github.com/JueonPark/sym/issues/144).

## Starting evidence

Audit against merged main `4e67545d2e6331802a7e295c65c135cf90b26a4f` / PR #140. The dialect currently defines transpose, reshape, pad, and layout plan results. The public exporter is layout-only, with wire v0 / manifest schema 1 / interface 1.

`Quant.h` currently specifies reciprocal-scale multiplication, clamp to [-128,127], round-to-nearest-even, and NaN→-128 for its zero-point-free int8 kernels. `convertF32F16` specifies RNE, preserved subnormals, overflow→infinity, and bit agreement only for non-NaN inputs. CUDA dequantization is signed-i8 times scale. These are audited kernel contracts, not evidence that arbitrary PyTorch quantization or nonzero zero points already match them.

## File map

| Action | Path | Responsibility |
| --- | --- | --- |
| Create | `docs/reloc-typed-semantics.md` | Normative numerical, parameter, and operation-order contract |
| Modify | `sym/dialect/reloc/IR/RelocOps.td`, `RelocOps.cpp` | Cast/quantize/dequantize definitions, assembly, builders, and verifiers |
| Modify | `sym/dialect/reloc/IR/RelocAttrs.td`, `RelocAttributes.cpp` | Shared numerical-policy and quantization-parameter attributes |
| Create | `test/dialect/reloc/typed_ops.mlir`, `typed_invalid_ops.mlir` | Valid parse/print and verifier diagnostics |
| Modify | `libreloc/test/QuantTest.cpp`, `CudaKernelsTest.cpp` | Focused characterization cases needed to distinguish current kernels from the declared contract |
| Consult | `libreloc/include/reloc/Quant.h`, `CudaKernels.h`, `libreloc/quant/`, `libreloc/cuda/CudaKernels.cu` | Existing CPU/CUDA behavior and eligibility restrictions |

## Task 1: Publish the numerical and parameter contract

**Interfaces:** Produce `docs/reloc-typed-semantics.md` as the shared authority for C2–C4/R3. Its tables specify each operation's legal type pair, arithmetic order, rounding, saturation, signedness, special values, and parameter interpretation.

- [ ] Write separate tables for f32→f16 narrowing, f16→f32 widening, f32→s8 quantization, and s8→f32 dequantization. Specify ties, signed zero, subnormals, overflow/underflow, infinities, and NaNs. State whether NaN payload/sign preservation is promised or explicitly outside conformance.
- [ ] Specify whether the quantization parameter denotes scale or reciprocal scale, and exactly when reciprocal formation, multiplication/division, zero-point addition, rounding, and saturation occur. Do not assume `x / scale` and `x * invScale` are bit equivalent. Include intermediate precision and floating-point-environment assumptions.
- [ ] Define per-tensor and per-channel forms. The channel axis is relative to the typed operation's logical input, before later layout folds; rank-one parameter length must equal that axis's extent. Define broadcast rules, valid zero-point range, finite strictly positive scales, and the treatment of symbolic channel lengths.
- [ ] Separate inline constants from declared runtime parameter operands. Record element type, shape, stable binding identity, and static-versus-bind-time validation obligations. No raw address is serialized as a parameter identity.
- [ ] Characterize existing CPU scalar/SIMD, CUDA, and pinned Torch behavior on the same vectors. Publish a compatibility table: directly usable, needs a new reference/adapter path in R3, or unsupported. Nonzero zero points and a different channel mapping must not be attributed to the existing kernels without evidence.

Use at least these witness classes in the contract and downstream tests:

```text
RNE ties at unit scale, zero point 0: [-2.5,-1.5,-0.5,0.5,1.5,2.5]
signed-i8 limits: [-129,-128,-127,126,127,128]
binary16 boundaries: ±0, 2^-24, 2^-14, 65504, finite overflow
specials: NaN, +infinity, -infinity
parameters: positive finite scale; 0, negative, NaN, infinity; zero point -128/127 and out-of-range values
channels: shape [2,3,4], axis 1, three distinct parameter entries
```

For the existing symmetric RNE kernel subset, the tie vector must produce `[-2,-2,0,0,2,2]`; any new policy must have its own explicitly different name and reference. The document must resolve each numerical policy before C2/C3 code relies on it.

## Task 2: Define operations and enforce static legality

**Interfaces:** Produce `reloc.cast`, `reloc.quantize`, and `reloc.dequantize`, with builders/verifiers and reusable policy/parameter attributes. Each consumes a `!sym.tensor` value and returns the same logical shape with the requested element type; quantization operations additionally carry the declared inline or runtime parameters.

- [ ] Add failing parse/print and diagnostic tests before implementing the operations. Use generic MLIR operation syntax initially if custom syntax is introduced in the same task.
- [ ] Implement deterministic assembly and builder behavior. Roundtrips must preserve exact inline parameter bits, signedness, channel axis, operand/binding order, and numerical policy.
- [ ] Reject unsupported dtype pairs, changed shape/rank, unknown policies, invalid constant scales/zero points, invalid axes, incompatible parameter ranks, and statically wrong channel lengths. Signed-i8 semantics must remain explicit even when tensor storage uses MLIR signless `i8`.
- [ ] Preserve symbolic extents and runtime parameter declarations. Unknown runtime values are not validated by inspecting a capture hint; C3 must receive the remaining value/length guards.
- [ ] Test constant versus runtime parameters, per-tensor versus per-channel forms, a negative/out-of-range axis policy, rank mismatch, missing parameters, and duplicate/conflicting parameter declarations.
- [ ] Run the new verifier tests from failing to passing, then the existing dialect suite. Commit the definitions and their tests together.

## Task 3: Publish the compiler/runtime handoff without enabling execution

**Interfaces:** C2 receives the operation/policy schema and channel-axis meaning; C3 receives the parameter declaration and validation rules; C4/R3 receive numerical vectors and the compatibility table.

- [ ] Document required stage input/output types and fill-value domains so C2 cannot move a conversion across pad merely because element counts agree.
- [ ] Add a public-export regression showing that a verified typed operation remains a declared unsupported export until C2/C3 provide its complete typed representation and encoder. A valid MLIR module is not yet an executable typed artifact.
- [ ] Keep the existing v0 layout-only corpus and exporter behavior intact. Do not relax T2's same-dtype artifact checks or enable a typed frontend row in C1.
- [ ] Link every claimed numerical behavior to its normative table and characterization case, and list the concrete R3 kernel/reference gaps. Commit the semantic handoff with the verifier work.

## Acceptance and handoff

- [ ] All four initial value-conversion directions have explicit semantics, including non-finite inputs and rounding boundaries.
- [ ] Invalid type/shape/policy/constant-parameter combinations fail verification; dynamic obligations remain explicit for C3.
- [ ] Per-channel axis and parameter-length meaning is stable under later folding.
- [ ] Existing kernel/Torch differences are reconciled or explicitly excluded, including reciprocal-scale rounding and NaN behavior.
- [ ] C2–C4/R3 share one contract; C1 makes no unsupported execution or v1 artifact claim.

From a configured LLVM/MLIR 21.1.8 build:

```bash
export SYM_BUILD="$PWD/build/torch-cpu"
cmake --build "$SYM_BUILD" -j2
"$SYM_BUILD/sym/tools/sym-opt" test/dialect/reloc/typed_ops.mlir
"$SYM_BUILD/sym/tools/sym-opt" test/dialect/reloc/typed_invalid_ops.mlir -verify-diagnostics -split-input-file
cmake --build "$SYM_BUILD" --target check-sym
ctest --test-dir "$SYM_BUILD" --output-on-failure
```

Record CPU and actual CUDA characterization separately; skipped GPU tests do not qualify a CUDA semantic claim.
